<launch>
    <!-- Set use_sim_time true to generate trajectory bag file in same time domain -->
    <param name="/use_sim_time" value="true"/>
    
    <!-- Path of a rosbag file -->
    <arg name="bag_path" default="$(find path_recorder)/../bag/kitti_2011_09_30_drive_0027_synced"/>
    <!-- ROS paramaters -->
    <arg name="visualize" default="true"/>
    <arg name="odom_frame" default="odom"/>
    <arg name="base_frame" default="velodyne"/>
    <arg name="topic" default="/velodyne_points"/>
    <arg name="publish_odom_tf" default="true"/>
    <arg name="record" default="false"/>

    <!-- KISS-ICP paramaters -->
    <arg name="deskew" default="false"/>
    <arg name="max_range" default="100.0" />
    <arg name="min_range" default="1.0" />
    <arg name="voxel_size" default="" />
    <!-- Odometry node -->
  
    <!-- Run KISS-ICP --> 
    <group>
        <node pkg="kiss_icp" type="odometry_node" name="odometry_node" output="screen">
            <!-- ROS params -->
            <remap from="pointcloud_topic" to="$(arg topic)"/>
            <param name="odom_frame" value="$(arg odom_frame)"/>
            <param name="base_frame" value="$(arg base_frame)"/>
            <param name="publish_odom_tf" value="$(arg publish_odom_tf)"/>
            <param name="visualize" value="$(arg visualize)"/>
            <!-- KISS-ICP params -->
            <param name="max_range" value="$(arg max_range)"/>
            <param name="min_range" value="$(arg min_range)"/>
            <param name="deskew" value="$(arg deskew)"/>
            <param name="voxel_size" value="$(arg voxel_size)"/>
            <param name="max_points_per_voxel" value="20"/>
            <param name="initial_threshold" value="2.0"/>
            <param name="min_motion_th" value="0.1" />
        </node>
        
        <arg name="rviz" default="true" />
        <group if="$(arg rviz)">
            <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find kiss_icp)/rviz/kiss_icp_ros1.rviz"/>
            <!-- <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz" args="-d $(find aloam_velodyne)/rviz_cfg/aloam_velodyne.rviz" /> -->
        </group>
    </group>


	<!-- Remap point cloud topic -->
	<remap from="points_raw" to="velodyne_points" />
	<!-- Play kitti rosbag file -->
    <node pkg="rosbag" type="play" name="kitti" args="--clock $(arg bag_path)/kitti.bag" output="log" />

	<!-- Launch path_recorder to save a trajectory from KISS-ICP -->	
    <node pkg="path_recorder" type="path_recorder" name="kiss_recorder" output="screen">
        <param name="algorithm_name" type="string" value="kiss_icp" />
        <param name="path_topic" type="string" value="/kiss/trajectory" />
        <param name="save_to" type="string" value="$(arg bag_path)" />
    </node>
</launch>
